src/dmd/console.d

```diff
@@ -37,216 +51,195 @@ enum Color : int
+ *  6: fast blink
+ *  7: reverse video
+ *  8: hidden
+ */
+private final class ANSIConsole : Console
+{
+  nothrow:
 
-        @property FILE* fp() { return _fp; }
-        /**
-         Tries to detect whether DMD has been invoked from a terminal.
-         Returns: `true` if a terminal has been detect, `false` otherwise
-         */
-        static bool detectTerminal()
-        {
-            import core.stdc.stdlib : getenv;
-            const(char)* term = getenv("TERM");
-            import core.stdc.string : strcmp;
-            return isatty(STDERR_FILENO) && term && term[0] && strcmp(term, "dumb") != 0;
-        }
+  private:
+    FILE* _fp;
 
-        static Console* create(FILE* fp)
-        {
-            auto c = new Console();
-            c._fp = fp;
-            return c;
-        }
+  public:
 
-        void setColorBright(bool bright)
-        {
-            fprintf(_fp, "\033[%dm", bright);
-        }
+    this(FILE* fp) { _fp = fp; }
 
-        void setColor(Color color)
-        {
-            fprintf(_fp, "\033[%d;%dm", color & Color.bright ? 1 : 0, 30 + (color & ~Color.bright));
-        }
+    @property FILE* fp() { return _fp; }
 
-        void resetColor()
-        {
-            fputs("\033[m", _fp);
-        }
+    void setColorBright(bool bright)
+    {
+        fprintf(_fp, "\033[%dm", bright);
     }
-    else
+
+    void setColor(Color color)
     {
-        @property FILE* fp() { assert(0); }
+        fprintf(_fp, "\033[%d;%dm", color & Color.bright ? 1 : 0, 30 + (color & ~Color.bright));
+    }
 
-        static Console* create(FILE* fp)
-        {
-            return null;
-        }
+    void resetColor()
+    {
+        fputs("\033[m", _fp);
+    }
+}
 
-        void setColorBright(bool bright)
-        {
-            assert(0);
-        }
+/**
+ Tries to detect whether DMD has been invoked from a terminal.
+ Returns: `true` if a terminal has been detected, `false` otherwise
+ */
+bool detectTerminal() nothrow
+{
+    version (Posix)
+    {
+        import core.stdc.stdlib : getenv;
+        const(char)* term = getenv("TERM");
+        import core.stdc.string : strcmp;
+        return isatty(STDERR_FILENO) && term && term[0] && strcmp(term, "dumb") != 0;
+    }
+    else version (Windows)
+    {
+        auto h = WindowsConsole.getStdHandle(stderr);
+        if (h is null)
+            return false;
 
-        void setColor(Color color)
-        {
-            assert(0);
-        }
+        CONSOLE_SCREEN_BUFFER_INFO sbi;
+        return GetConsoleScreenBufferInfo(h, &sbi) != 0;
+    }
+}
 
-        void resetColor()
-        {
-            assert(0);
-        }
+/**
+ * Creates an instance of Console connected to stream fp.
+ * Params:
+ *      fp = io stream
+ * Returns:
+ *      reference to created Console
+ */
+Console createConsole(FILE* fp) nothrow
+{
+    version (Windows)
+    {
+        if (auto c = WindowsConsole.create(fp))
+            return c;
     }
 
+    return new ANSIConsole(fp);
```

Comment:
This ANSI fallback on Windows is the 'no-op change' - Windows-specific `Console.create()` would previously return null if no terminal was detected, disabling fancy colors.